vdaf/prio3/prio3_test.go

```diff
@@ -0,0 +1,234 @@
+	got := testPrio3(t, s, input)
+	test.CheckOk(got != nil, "prio3 sumvec failed", t)
+	if !slices.Equal(*got, want) {
+		test.ReportError(t, *got, want)
+	}
+}
+
+func TestHistogram(t *testing.T) {
+	input := []uint64{2, 1, 1, 2, 0, 3, 3, 3, 3, 3}
+	want := []uint64{
+		1, // how many zeros
+		2, // how many ones
+		2, // how many twos
+		5, // how many threes
+	}
+	h, err := histogram.New(NumShares, 4, 3, Context)
+	test.CheckNoErr(t, err, "new histogram failed")
+	got := testPrio3(t, h, input)
+	test.CheckOk(got != nil, "prio3 histogram failed", t)
+	if !slices.Equal(*got, want) {
+		test.ReportError(t, *got, want)
+	}
+}
+
+func TestMultiHotCountVec(t *testing.T) {
+	const MaxWeight = 2
+	input := [][]bool{
+		{false, false, false, false, false}, // Hamming weight = 0 <= MaxWeight
+		{false, false, false, false, true},  // Hamming weight = 1 <= MaxWeight
+		{true, false, false, false, true},   // Hamming weight = 2 <= MaxWeight
+	}
+	// Element-wise count of binary vectors of weight at most MaxWeight.
+	want := []uint64{1, 0, 0, 0, 2}
+	m, err := mhcv.New(NumShares, 5, MaxWeight, 3, Context)
+	test.CheckNoErr(t, err, "new MultiHotCountVec failed")
+	got := testPrio3(t, m, input)
+	test.CheckOk(got != nil, "prio3 multiHotCountVec failed", t)
+	if !slices.Equal(*got, want) {
+		test.ReportError(t, *got, want)
+	}
+}
+
+func testPrio3[
+	P Prio3[
+		Measurement, Aggregate,
+		AggShare, InputShare, OutShare, PrepShare, PrepState,
+	],
+	Measurement, Aggregate any,
+	AggShare, InputShare, OutShare, PrepShare, PrepState any,
+](t testing.TB, p P, measurements []Measurement) *Aggregate {
+	params := p.Params()
+	shares := params.Shares()
+	var verifyKey prio3.VerifyKey
+	_, err := io.ReadFull(rand.Reader, verifyKey[:])
+	test.CheckNoErr(t, err, "read verifyKey failed")
+
+	aggShares := make([]AggShare, shares)
+	for i := range aggShares {
+		aggShares[i] = p.AggregationInit()
+	}
+
+	for _, mi := range measurements {
+		var nonce prio3.Nonce
+		_, err = io.ReadFull(rand.Reader, nonce[:])
+		test.CheckNoErr(t, err, "read nonce failed")
+
+		randb := make([]byte, params.RandSize())
+		_, err = io.ReadFull(rand.Reader, randb)
+		test.CheckNoErr(t, err, "read rand bytes failed")
+
+		var pubShare prio3.PublicShare
+		var inputShares []InputShare
+		pubShare, inputShares, err = p.Shard(mi, &nonce, randb)
+		test.CheckNoErr(t, err, "shard failed")
+		testMarshal(t, &pubShare, &params)
+		for i := range inputShares {
+			testMarshal(t, &inputShares[i], &params, uint(i))
+		}
+
+		var prepStates []*PrepState
+		var outboundPrepShares []PrepShare
+		for i := range shares {
+			state, share, errx := p.PrepInit(
+				&verifyKey, &nonce, i, pubShare, inputShares[i])
+			test.CheckNoErr(t, errx, "prepare init failed")
+
+			prepStates = append(prepStates, state)
+			outboundPrepShares = append(outboundPrepShares, *share)
+		}
+
+		testMarshal(t, prepStates[0], &params)
+		testMarshal(t, &outboundPrepShares[0], &params)
+
```

Comment:
It's worth testing all prep state/shares and not just the first, as they might have different formats depending on whether the aggregator is the leader or a helper.

Reply:
done